File: pipelines/flink-jobs/src/main/java/com/ainsoft/brain/flink/jobs/aggregate/WindowFeatureProcessFunction.java
package com.ainsoft.brain.flink.jobs.aggregate;

import com.ainsoft.brain.core.events.WindowFeature;
import org.apache.flink.streaming.api.functions.windowing.ProcessWindowFunction;
import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
import org.apache.flink.util.Collector;

import java.lang.Iterable;

public final class WindowFeatureProcessFunction
    extends ProcessWindowFunction<SumCount, WindowFeature, SensorKey, TimeWindow> {

  private final long windowSizeMs;

  public WindowFeatureProcessFunction(long windowSizeMs) {
    this.windowSizeMs = windowSizeMs;
  }

  @Override
  public void process(
      SensorKey key,
      Context context,
      Iterable<SumCount> elements,
      Collector<WindowFeature> out
  ) {
    SumCount agg = elements.iterator().next();
    double mean = agg.count() == 0L ? 0.0 : agg.sum() / (double) agg.count();
    out.collect(
        new WindowFeature(
            key.deviceId(),
            key.sensorType(),
            context.window().getStart(),
            context.window().getEnd(),
            windowSizeMs,
            agg.count(),
            mean,
            "v1"
        )
    );
  }
}

File: pipelines/flink-jobs/src/main/scala/com/ainsoft/brain/flink/jobs/aggregate/WindowFeatureProcessFunction.scala
package com.ainsoft.brain.flink.jobs.aggregate

import com.ainsoft.brain.core.events.WindowFeature
import org.apache.flink.streaming.api.functions.windowing.ProcessWindowFunction
import org.apache.flink.streaming.api.windowing.windows.TimeWindow
import org.apache.flink.util.Collector

final class WindowFeatureProcessFunction(windowSizeMs: Long)
    extends ProcessWindowFunction[SumCount, WindowFeature, SensorKey, TimeWindow] {

  override def process(
      key: SensorKey,
      context: ProcessWindowFunction[SumCount, WindowFeature, SensorKey, TimeWindow]#Context,
      elements: java.lang.Iterable[SumCount],
      out: Collector[WindowFeature]
  ): Unit = {
    val it = elements.iterator()
    if (it.hasNext) {
      val agg = it.next()
      val mean = if (agg.count == 0L) 0.0 else agg.sum / agg.count.toDouble
      out.collect(
        WindowFeature(
          deviceId = key.deviceId,
          sensorType = key.sensorType,
          windowStartMs = context.window.getStart,
          windowEndMs = context.window.getEnd,
          windowSizeMs = windowSizeMs,
          count = agg.count,
          meanValue = mean,
          featureSchemaVersion = "v1"
        )
      )
    }
  }
}
